_WKContextMenuElementInfo should expose a hasEntireImage property

jeffmapple opened the pull request:
#### 9643b29c75f08b2aed6b8abd2ae1b45f73e42015
<pre>
_WKContextMenuElementInfo should expose a hasEntireImage property
<a href="https://bugs.webkit.org/show_bug.cgi?id=253258">https://bugs.webkit.org/show_bug.cgi?id=253258</a>
&lt;rdar://106152692&gt;

Reviewed by Aditya Keerthi.

Add a hasEntireImage property to _WKContextMenuElementInfo. This property is true if we&apos;re showing a
context menu for an image and the image has been loaded.

As part of this I moved hasLocalDataForURL() from WebKit::WebPage to WebCore::Page so we can can use
it from HitTestResult::hasEntireImage() in WebCore. I also changed
API::ContextMenuElementInfoMac::create() to take a single const WebKit::ContextMenuContextData&amp;
parameter rather than adding a third parameter since all of the information needed to create this
class is derived from ContextMenuContextData.

* Source/WebCore/page/ContextMenuContext.cpp:
(WebCore::ContextMenuContext::ContextMenuContext):
* Source/WebCore/page/ContextMenuContext.h:
(WebCore::ContextMenuContext::hasEntireImage const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::hasLocalDataForURL):
Moved from WebKit::WebPage.

* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasEntireImage const):
Returns true if the context menu is for an image, the image rectangle isn&apos;t empty, and the page has
loaded the image data.

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/ContextMenuContextData.cpp:
(WebKit::ContextMenuContextData::encode const):
(WebKit::ContextMenuContextData::decode):
* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::hasEntireImage const):
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo hasEntireImage]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageHasLocalDataForURL):
Call hasLocalDataForURL() on the WebCore::Page instead of the WebKit::WebPage. This was the only
caller of this function.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasLocalDataForURL): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added a new ContextMenuElementInfoHasEntireImage test.

Canonical link: <a href="https://commits.webkit.org/261793@main">https://commits.webkit.org/261793@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d306002167eb98aa4a5d8109d5ad6eb4a117813

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121332 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23326 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13148 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5770 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100580 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105912 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99292 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1119 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46344 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14284 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1159 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14984 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10503 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20301 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53152 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8226 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16833 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -240,7 +240,7 @@
         return completionHandler(menu);
 
-    auto contextMenuElementInfo = API::ContextMenuElementInfoMac::create(data.webHitTestResultData().value(), data.qrCodePayloadString());
+    auto contextMenuElementInfo = API::ContextMenuElementInfoMac::create(data.webHitTestResultData().value(), data.qrCodePayloadString(), data.hasEntireImage());
```
At some point we should make the `ContextMenuElementInfoMac` just take a `const ContextMenuContextData&` parameter.

Not something that needs to be done in this change though.

jeffmapple replied:
It did do that until recently, when you changed it in https://commits.webkit.org/260897@main. 😆

pxlcoder replied:
It used to take a `WebHitTestResultData` before, not a `ContextMenuContextData` 😅

jeffmapple replied:
Strike that, I misremembered what your change did. This is a good idea, since I'll likely need to expose more properties from `ContextMenuElementInfoMac` in the future. I'll do that when I make those additional changes.

jeffmapple replied:
Actually, since I'm going to make the other change you suggested now, I might as well do this now too.

jeffmapple replied:
I did this in the commit I just uploaded.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebContextMenu.cpp:
```diff
@@ -63,10 +63,15 @@ void WebContextMenu::show()
 
+    auto imageURL = hitTestResult.absoluteImageURL();
+    if (!imageURL.isEmpty() && !hitTestResult.imageRect().isEmpty() && m_page->hasLocalDataForURL(imageURL))
+        contextMenuContextData.setHasEntireImage(true);
```
I think it might make more sense to move this logic into the `ContextMenuContextData` constructor, which has access to the `HitTestResult`.

jeffmapple replied:
I considered that, but `ContextMenuContextData` doesn't have access to the page.

pxlcoder replied:
You could get it from the `HitTestResult`, `innerNodeFrame()->page()`.

jeffmapple replied:
That's actually the `WebCore::Page`, and  `hasLocalDataForURL()` is a `WebKit::WebPage` function. But I still think you're right that the logic makes more sense in `ContextMenuContextData`, so I'm going to add a `WebPage&` to the constructor.

pxlcoder replied:
I think that would be incorrect since `ContextMenuContextData` is in `WebKit/Shared`, but `WebPage` is a `WebProcess` only concept.

How about we move the implementation of `hasLocalDataForURL()` to `WebCore::Page` instead?

jeffmapple replied:
That sounds reasonable. Assuming I can do that, do you think it also makes sense to move the calculation of `hasEntireImage` into `WebCore::ContextMenuContext` as well and just have `WebKit::ContextMenuData` initialize its member variable from `ContextMenuContext`?

pxlcoder replied:
Sounds good to me!

pxlcoder replied:
Actually, now that I think about it, we could just put everything on `HitTestResult` / `WebHitTestResultData` / `_WKHitTestResult`, since this just operates on hit test result data.

The reason the QR code payload is a separate property is that it cannot be determined at hit test result time.

pxlcoder replied:
@whsieh @hortont424 Any thoughts on whether this SPI should live on `_WKHitTestResult` or `_WKContextMenuElementInfo`? Today, it will only be used for context menus, but in reality it's simply derived from properties on `_WKHitTestResult` and could be generally useful.

jeffmapple replied:
I left the property on `_WKContextMenuElementInfo` for now, but I'll leave this conversation open to get other people's opinions.

pxlcoder commented on Source/WebCore/page/ContextMenuContext.cpp:
```diff
@@ -42,7 +44,18 @@ ContextMenuContext::ContextMenuContext(Type type, const HitTestResult& hitTestRe
+        return;
+
+    if (auto page = innerNodeFrame->page())
+        m_hasEntireImage = page->hasLocalDataForURL(imageURL);
```
I think this would be better encapsulated as a `const` function on `WebCore::HitTestResult`. Then, we could have `m_hasEntireImage(hitTestResult.hasEntireImage())` in the constructor.

Not a big fan of the early returns in the constructor, since we may have to refactor if we were to add more properties.

jeffmapple replied:
Excellent point, will fix.

jeffmapple committed:
_WKContextMenuElementInfo should expose a hasEntireImage property
https://bugs.webkit.org/show_bug.cgi?id=253258
<rdar://106152692>

Reviewed by Aditya Keerthi.

Add a hasEntireImage property to _WKContextMenuElementInfo. This property is true if we're showing a
context menu for an image and the image has been loaded.

As part of this I moved hasLocalDataForURL() from WebKit::WebPage to WebCore::Page so we can can use
it from HitTestResult::hasEntireImage() in WebCore. I also changed
API::ContextMenuElementInfoMac::create() to take a single const WebKit::ContextMenuContextData&
parameter rather than adding a third parameter since all of the information needed to create this
class is derived from ContextMenuContextData.

* Source/WebCore/page/ContextMenuContext.cpp:
(WebCore::ContextMenuContext::ContextMenuContext):
* Source/WebCore/page/ContextMenuContext.h:
(WebCore::ContextMenuContext::hasEntireImage const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::hasLocalDataForURL):
Moved from WebKit::WebPage.

* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasEntireImage const):
Returns true if the context menu is for an image, the image rectangle isn't empty, and the page has
loaded the image data.

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/ContextMenuContextData.cpp:
(WebKit::ContextMenuContextData::encode const):
(WebKit::ContextMenuContextData::decode):
* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::hasEntireImage const):
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo hasEntireImage]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageHasLocalDataForURL):
Call hasLocalDataForURL() on the WebCore::Page instead of the WebKit::WebPage. This was the only
caller of this function.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasLocalDataForURL): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added a new ContextMenuElementInfoHasEntireImage test.

Canonical link: https://commits.webkit.org/261793@main
